### PR TITLE
Fix the rewriting of the mongo config on agent start.

### PR DESCRIPTION
**Pull request by howbazaar**
During the 2.6 refactor of the mongo service code, the rewriting of the mongod service config was mistakenly removed.

This branch reinstates the 2.5 logic around the checking of the service. Also adds a test and makes sure that the arguments for the config are output in a deterministic order.

## QA steps

```
juju bootstrap lxd test
juju model-config -m controller logging-config=juju=info:juju.mongo=debug
juju run -m controller --machine 0 'cat /lib/systemd/system/juju-db/juju-db.service'
# Note no --wiredTigerCacheSizeGB flag
juju controller-config mongo-memory-profile=low
# Wait a few seconds
juju run -m controller --machine 0 'cat /lib/systemd/system/juju-db/juju-db.service'
# Note that the flag turned up.
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1840731

### CODE REVIEW

File: mongo/mongo.go [hunk 1]
```diff
@@ -582,20 +582,44 @@ func ensureServer(args EnsureServerParams, mongoKernelTweaks map[string]string)
...
+	// Installed tells us if there exists a service of the right name.
+	installed, err := svc.Installed()
 	if err != nil {
 		return zeroVersion, errors.Trace(err)
 	}
-	if running {
-		return mongodVersion, nil
+	if installed {
+		// Exists does a check against the contents of the service config file.
+		// The return value is true iff the contest is the same.
```
Comment (jtroup) [inline mongo/mongo.go]: Typo: s/contest/content/

File: mongo/mongo.go [hunk 2]
```diff
@@ -582,20 +582,44 @@ func ensureServer(args EnsureServerParams, mongoKernelTweaks map[string]string)
...
-	running, err := svc.Running()
+	// Installed tells us if there exists a service of the right name.
+	installed, err := svc.Installed()
 	if err != nil {
 		return zeroVersion, errors.Trace(err)
 	}
-	if running {
-		return mongodVersion, nil
+	if installed {
+		// Exists does a check against the contents of the service config file.
```
Comment (jtroup) [inline mongo/mongo.go]: It might be easier to read as: 'Exists()' rather than just 'Exists'.

File: mongo/mongo.go (hunk 1 above)
Comment (howbazaar) [reply mongo/mongo.go]: Updated.

File: mongo/mongo.go (hunk 2 above)
Comment (howbazaar) [reply mongo/mongo.go]: Seems reasonable to me. Updated.

File: mongo/mongo.go [hunk 3]
```diff
@@ -582,20 +582,44 @@ func ensureServer(args EnsureServerParams, mongoKernelTweaks map[string]string)
...
 
+	// We want to write or rewrite the contents of the service.
+	// Stop is a no-op if the service doesn't exist or isn't running.
+	if err := svc.Stop(); err != nil {
+		return zeroVersion, errors.Annotatef(err, "failed to stop mongo")
+	}
 	dbDir := DbDir(args.DataDir)
 	if err := makeJournalDirs(dbDir); err != nil {
-		return zeroVersion, errors.Errorf("error creating journal directories: %v", err)
+		return zeroVersion, fmt.Errorf("error creating journal directories: %v", err)
```
Comment (wallyworld) [inline mongo/mongo.go]: why this change?

Comment (howbazaar) [reply mongo/mongo.go]: Because I copied the block of code from 2.5 branch. Will update.

**Comment by howbazaar** [comment]
$$merge$$

**Comment by howbazaar** [comment]
$$merge$$ intermittent failure, relationunit_test.go:243: relationUnitSuite.TestWatchRelationUnits
